File: lib/node/mango.js
const net = require("net");

const socketPath = () => {
  const sig = process.env.MANGO_INSTANCE_SIGNATURE;
  if (sig) return sig;
  throw new Error("MANGO_INSTANCE_SIGNATURE is not set");
};

const call = (cmd) =>
  new Promise((resolve, reject) => {
    const sock = net.createConnection(socketPath());
    let buf = "";
    let done = false;
    const finish = (err, value) => {
      if (done) return;
      done = true;
      sock.destroy();
      if (err) reject(err);
      else resolve(value);
    };
    sock.on("error", (e) => finish(e));
    sock.on("connect", () => sock.write(cmd + "\n"));
    sock.on("data", (chunk) => {
      buf += chunk.toString();
      const nl = buf.indexOf("\n");
      if (nl === -1) return;
      try {
        finish(null, JSON.parse(buf.slice(0, nl)));
      } catch (e) {
        finish(e);
      }
    });
    sock.on("end", () => {
      if (!done) finish(new Error("connection closed without reply: " + cmd));
    });
  });

const ensure = (reply) => {
  if (reply.error) throw new Error(reply.error);
  return reply;
};

const retry = async (cmd, attempts = 50, delay = 100) => {
  for (let i = 0; ; i++) {
    try {
      return ensure(await call(cmd));
    } catch (e) {
      if (i === attempts - 1) throw e;
      await new Promise((r) => setTimeout(r, delay));
    }
  }
};

const get = async (...spec) => ensure(await call("get " + spec.join(" ")));

const setOption = async (key, value) =>
  ensure(await call("setoption " + key + " " + value));

const dispatch = async (fn, ...args) =>
  ensure(await call("dispatch " + [fn, ...args].join(",")));

const unsetBind = async (mode, mods, keysym, family = "bind") =>
  ensure(await call(`unset bind ${mode} ${mods} ${keysym} ${family}`));

const unsetRule = async (kind, spec) =>
  ensure(await call(`unset ${kind} ${spec}`));

const watch = async (subject, onEvent) => {
  const sock = net.createConnection(socketPath(), () => {
    if (sock.destroyed) return;
    sock.write("watch " + subject + "\n");
  });
  sock.setEncoding("utf8");
  let buf = "";
  for await (const chunk of sock) {
    buf += chunk;
    let nl;
    while ((nl = buf.indexOf("\n")) !== -1) {
      const line = buf.slice(0, nl);
      buf = buf.slice(nl + 1);
      if (line) onEvent(ensure(JSON.parse(line)));
    }
  }
};

module.exports = {
  socketPath,
  call,
  retry,
  get,
  setOption,
  dispatch,
  unsetBind,
  unsetRule,
  watch,
};

File: lib/lua/mango.lua
local unix = require("socket.unix")
local socket = require("socket")
local json = require("cjson")
local M = {}

local buffers = {}

local function socket_path()
  local sig = os.getenv("MANGO_INSTANCE_SIGNATURE")
  if sig then return sig end
  error("MANGO_INSTANCE_SIGNATURE is not set")
end

function M.connect()
  local sock, err = unix()
  if not sock then error(err) end
  sock:settimeout(10)
  local ok, cerr = sock:connect(socket_path())
  if not ok then error(cerr) end
  return sock
end

function M.call(cmd)
  local sock = M.connect()
  local ok, err = sock:send(cmd .. "\n")
  if not ok then sock:close() error(err) end
  local line, rerr = sock:receive("*l")
  sock:close()
  if not line then error(rerr or ("connection closed without reply: " .. cmd)) end
  return json.decode(line)
end

function M.ensure_ok(reply)
  if reply.error then error(reply.error) end
  return reply
end

function M.retry(cmd, attempts, delay)
  attempts = attempts or 50
  delay = delay or 0.1
  for _ = 1, attempts do
    local ok, reply = pcall(M.call, cmd)
    if ok then return M.ensure_ok(reply) end
    socket.sleep(delay)
  end
  error("mango not reachable after " .. attempts .. " attempts: " .. cmd)
end

function M.get(...)
  return M.ensure_ok(M.call("get " .. table.concat({ ... }, " ")))
end

function M.set_option(key, value)
  M.ensure_ok(M.call("setoption " .. key .. " " .. value))
end

function M.dispatch(function_name, ...)
  M.ensure_ok(M.call("dispatch " .. table.concat({ function_name, ... }, ",")))
end

function M.unset_bind(mode, mods, keysym, family)
  family = family or "bind"
  M.ensure_ok(M.call("unset bind " .. mode .. " " .. mods .. " " .. keysym .. " " .. family))
end

function M.unset_rule(kind, spec)
  M.ensure_ok(M.call("unset " .. kind .. " " .. spec))
end

function M.watch(subject, on_event)
  local sock = M.connect()
  local ok, err = sock:send("watch " .. subject .. "\n")
  if not ok then sock:close() error(err) end
  sock:settimeout(nil)
  while true do
    local line, rerr = sock:receive("*l")
    if not line then break end
    if line ~= "" then
      on_event(M.ensure_ok(json.decode(line)))
    end
  end
  sock:close()
end

local function next_line(sock)
  while true do
    local line, err, partial = sock:receive("*l")
    if line then
      local head = buffers[sock]
      buffers[sock] = nil
      if head and #head > 0 then return head .. line end
      return line
    end
    if err == "timeout" then
      local head = buffers[sock] or ""
      buffers[sock] = head .. (partial or "")
      return nil
    end
    return nil, err
  end
end

function M.open_watch(subject)
  local sock = M.connect()
  local ok, err = sock:send("watch " .. subject .. "\n")
  if not ok then sock:close() error(err) end
  sock:settimeout(0)
  return sock
end

function M.read_event(sock)
  while true do
    local line, err = next_line(sock)
    if not line then return nil, err end
    if line ~= "" then
      local ok, obj = pcall(json.decode, line)
      if ok and not obj.error then return obj end
    end
  end
end

return M